includes/class-wc-payments-order-success-page.php

```diff
@@ -255,6 +258,59 @@ private function format_addtional_thankyou_order_received_text( string $addition
 		return sprintf( '<div class="woocommerce-info">%s</div>', $additional_text );
 	}
 
+	/**
+	 * Replace the order received text with a failure message when the order status is 'failed'.
+	 *
+	 * @param string $text The original thank you text.
+	 * @return string
+	 */
+	public function replace_order_received_text_for_failed_orders( $text ) {
+		global $wp;
+
+		$order_id = absint( $wp->query_vars['order-received'] );
+		$order    = wc_get_order( $order_id );
+
+		if ( ! $order ) {
+			return $text;
+		}
+
+		$intent_id      = $order->get_meta( '_intent_id', true );
+		$payment_method = $order->get_payment_method();
+
+		// Strip the gateway ID prefix from the payment method.
+		$payment_method_type = str_replace( WC_Payment_Gateway_WCPay::GATEWAY_ID . '_', '', $payment_method );
+
+		$should_show_failure = false;
+
+		// Check order status first to avoid unnecessary API calls.
+		if ( $order->has_status( Order_Status::FAILED ) ) {
+			$should_show_failure = true;
+		} elseif ( ! empty( $intent_id ) && ! empty( $payment_method_type ) && in_array( $payment_method_type, Payment_Method::REDIRECT_PAYMENT_METHODS, true ) ) {
+			// For redirect-based payment methods that haven't been marked as failed yet, check the intent status.
+			// Add a small delay to allow the intent to be updated.
+			sleep( 1 );
+
+			$intent        = Get_Intention::create( $intent_id );
+			$intent        = $intent->send();
+			$intent_status = $intent->get_status();
+
```

Comment:
The order status for some payment methods, like WeChat Pay, is updated via AJAX and the order hasn't yet updated by the time the `woocommerce_thankyou_order_received_text` runs. Getting the error status from the intent seems to be the best way.

Reply:
This works also, but I'm not sure it's faster. I've tried it here: [c73fafa](https://github.com/Automattic/woocommerce-payments/pull/10554/commits/c73fafa66fbe2f49283f420d1810cf52cae3fb83). The number of retries seems arbitrary but 6 seemed to be the sweet spot in my local testing. Let me know what you think.

Reply:
Yes, it seems that the results are not very consistent, I'm also surprised it takes that long to update the order, in my case the 6 retries were not enough. 

I think we have two options here:
1. Poll less times, but if the order still needs payment maybe provide a custom message that says that the process hasn't completed and ask the user to refresh the page (or go a step further and move the polling to the frontend, although that would definitely increase the scope significantly).
2. Restore the get intent approach (but maybe keep the polling mechanism with less retries?)

Although I like not having to make requests to the server, I think #2  may give us more efficient results?

Reply:
If I understand, option 2 proposes to keep the current polling mechanism with, say 3 retries, then move to the get intent approach if polling fails to capture the status update? If so, that sounds fine to me.

Reply:
For option two I actually meant to remove the order status polling and only poll the intents with a few retries. As the order status has proven to be way slower (I think the slowness is because the order is updated by an AJAX call performed by the original checkout page, so that's an additional roundtrip besides all the additional processing done by the endpoint)

Reply:
Got it! We can go with that. Though I was seeing consistent results using a short delay before checking the intent. Were you getting different results and think polling is necessary?

Reply:
I thought of it more like a failsafe measure but we can remove it to keep things simple. In the future it may make sense to revisit option (1), so I think it's fine if for now we keep things straightforward, IMO.

Reply:
Sounds good. For now, this should only affect a single payment method and only in test mode. But option 1 sounds like a good path to take should we ever need to revisit it.